aggregator/election_mgr.go

```diff
@@ -0,0 +1,505 @@
+	// Open opens the election manager for a given shard set.
+	Open(shardSetID string) error
+
+	// ElectionState returns the election state.
+	ElectionState() ElectionState
+
+	// Resign stops the election and resigns from the ongoing campaign if any, thereby
+	// forcing the current instance to become a follower. If the provided context
+	// expires before resignation is complete, the context error is returned, and the
+	// election is restarted if necessary.
+	Resign(ctx context.Context) error
+
+	// Close the election manager.
+	Close() error
+}
+
+var (
+	errElectionManagerExpectNotOpen = errors.New("election manager is expected to be not open")
+	errElectionManagerExpectOpen    = errors.New("election manager is expected to be open")
+	errLeaderNotChanged             = errors.New("leader has not changed")
+)
+
+type electionManagerState int
+
+const (
+	electionManagerNotOpen electionManagerState = iota
+	electionManagerOpen
+	electionManagerResigning
+	electionManagerClosed
+)
+
+// ElectionState is the election state.
+type ElectionState int
+
+// A list of supported election states.
+const (
+	FollowerState ElectionState = iota
+	LeaderState
+)
+
+func (state ElectionState) String() string {
+	switch state {
+	case FollowerState:
+		return "follower"
+	case LeaderState:
+		return "leader"
+	default:
+		panic(fmt.Sprintf("unknown election state %v", int(state)))
+	}
+}
+
+// MarshalJSON returns state as the JSON encoding of state.
+func (state ElectionState) MarshalJSON() ([]byte, error) {
+	return json.Marshal(state.String())
+}
+
+// UnmarshalJSON unmarshals JSON-encoded data into state.
+func (state *ElectionState) UnmarshalJSON(data []byte) error {
+	var str string
+	err := json.Unmarshal(data, &str)
+	if err != nil {
+		return err
+	}
+	switch str {
+	case FollowerState.String():
+		*state = FollowerState
+	case LeaderState.String():
+		*state = LeaderState
+	default:
+		return fmt.Errorf("unexpected json-encoded state: %s", str)
+	}
+	return nil
+}
+
+type electionManagerMetrics struct {
+	campaignCreateErrors tally.Counter
+	campaignErrors       tally.Counter
+	leaderErrors         tally.Counter
+	leaderNotChanged     tally.Counter
+	resignErrors         tally.Counter
+	resignNotFollower    tally.Counter
+	followerToLeader     tally.Counter
+	leaderToFollower     tally.Counter
+	electionState        tally.Gauge
+}
+
+func newElectionManagerMetrics(scope tally.Scope) electionManagerMetrics {
+	return electionManagerMetrics{
+		campaignCreateErrors: scope.Counter("campaign-create-errors"),
+		campaignErrors:       scope.Counter("campaign-errors"),
+		leaderErrors:         scope.Counter("leader-errors"),
+		leaderNotChanged:     scope.Counter("leader-not-changed"),
+		resignErrors:         scope.Counter("resign-errors"),
+		resignNotFollower:    scope.Counter("resign-not-follower"),
+		followerToLeader:     scope.Counter("follower-to-leader"),
+		leaderToFollower:     scope.Counter("leader-to-follower"),
```

Comment:
These two are emitted when a follower to leader or a leader to follower transition happens, which are events and better captured as counters. There is a separate gauge `electionState` below for the purpose of determining what the current election state is and can be used to know who the leader is.